### ISSUE TITLE
Feature request: conda env create/export with sha256 (sha256 in conda list --explicit)

### ISSUE THREAD

**Issue by kiwi0fruit** [conda/conda#7882]
UPD Feature request: add sha256 support in `conda list --explicit`

----

If there is a need for self-maintaned and self-assured security for python installation then having sha256 hashes in environment yaml is needed. So a functionality like [pip hash checking mode](https://pip.pypa.io/en/stable/reference/pip_install/#hash-checking-mode) can solve the issue.

It would be good to have such in conda. And this would also provide a better repeatability of fixed-version environments (and remove problems like [this](https://github.com/conda/conda/issues/5693)).

As I understood from https://github.com/conda/conda/issues/4956 there is a problem with hases being changed over time for same builds. I guess this can be fixed via additional conda channels like defaults-archive and conda-forge-archive. I guess there would be a problem with ID on these extra channels (when you change hash many times without changing build ID) but I guess this can be solved somehow (like simply counting added builds -- this may require modifying channels API -- and conda app downloads them all and calculates hashes).

**Comment by kiwi0fruit** [comment]
Or it can be fixed by adding the second conda package format: .tar.bz2.tar (or something similar).

It stores inside the first the first package of the build together with updatable requirements file (or anything else that may be updated after bublishing the build), and with file with sha256 hash (so it the system would work faster).

**Comment by kiwi0fruit** [comment]
Or it may be done without changing conda package format:

First we should find out which parts of conda package are updated without updating the build - metadata for dependencies? timestamps?

So conda unpacks the package, clears timestamps and deps. Then tars or zips. Then calculates the sha256. So this the sha256 that would be in conda env export. And the sha256 that would be checked by conda during install.

**Comment by kiwi0fruit** [comment]
I was focused on situation when every package in yaml is provided with hash
(and it should be sha256 as in pip I guess).

So in this case conda simply installs everything and doesn't "think". Or it
doesn't install if it wasn't able to find all packages with right hashes.

I wonder if it is needed to have mixed mode that only some packages are
with hash. But I myself see no point in it. For example pip uses this all
or nothing approach and I guess  with a reason.

>

**Comment by kiwi0fruit** [comment]
> The current thinking is to use a .pax file containing two seperate tar.bz2s (or some faster to decompress format).

Or three or four? I was also thinking about it (based on 7z ability to compress such archives that differ only in metadata to size of single archive). But why? Why not have only first tar.bz2 and updated metadata in pax or whatever?

**Comment by mingwandroid** [comment]
Because being able to verify that the core program code did not change is important from a security perspective and splitting like this makes metadata changes much faster to perform.

**Comment by kiwi0fruit** [comment]
For this the first tar.bz2 is used that is never changed. Why have the
second if we can have metadata itself only? Maybe it's for compatibility
and ease of integration: the pax has the first and the last version of the
build? The last is used in default mode and the fist in the hashed mode? If
we have two versions of tar.bz2 it's simpler to fix conda code than it we
have original tar.bz+updated metadata. Am I right?

вс, 21 окт. 2018, 16:43 Ray Donnelly <notifications@github.com>:

**Comment by mingwandroid** [comment]
Our/my current thinking is to have a .conda.pax format and *inside* that there are a few tarballs (not alongside as I think you are implying), and maybe even a few loose files for things on conda's hot-path so we can avoid decompression in that, for example `info/index.json` `info/files` and for conda-build `info/run_exports`.

Nothing's set in stone yet though.

**Comment by kiwi0fruit** [comment]
I also implied that tar.bz2 are inside the pax. I asked why have two
tar.bz2 (first and last?) when you can have only first + loose file with
updated metadata (that is again inside the pax)?

>

**Comment by kiwi0fruit** [comment]
Now I get it. Sounds reasonable. One last thing: these two tar.bz2's. Are they both conda packages? Or only one of them is a conda package and the second one is patch for it's metadata and tests? If the first option then the archive format should be capable of utilizing similar part of the conda packages to make the resulting archive not twice bigger.

**Comment by kiwi0fruit** [comment]
@mingwandroid By the way. If we are to imagine temporary solution based on the following idea:

> Or it may be done without changing conda package format:
> 
> First we should find out which parts of conda packages are updated without updating the build - metadata for dependencies? timestamps?
> 
> So conda unpacks the package, clears timestamps and deps. Then tars or zips. Then calculates the sha256. So this the sha256 that would be in conda env export. And the sha256 that would be checked by conda during install.

And imagine adding it to the tool like poetry. What parts of the conda package are being changed and what parts are contact within a build over time? So what part of the package can be used to calculate the hash?

**Comment by mingwandroid** [comment]
You just create two tarballs (
tar.zst I hope). One with so the actual package contents and the other with some the metadata. Those get hashed and stored in an outer .conda.pax file (which is the only thing to consider as a conda package the contents are implementation details)
 along with the test of the metadata.

I already specified what I think there rest should be earlier..

**Comment by kiwi0fruit** [comment]
Great. Looks like the best approach possible. Do you have any estimate on time when it may be implemented (even with a great range or uncertainty)?

**Comment by kiwi0fruit** [comment]
Actually the simplest solution to this problem is to stop changing packages "starting from today and on".

**Comment by kiwi0fruit** [comment]
@kalefranz Could you please elaborate if there are plans to add sha256 support to `conda list --explicit`.  If there is a need for self-maintaned and self-assured security for python installation then having sha256 hashes in environment yaml is needed. So a functionality like [pip hash checking mode](https://pip.pypa.io/en/stable/reference/pip_install/#hash-checking-mode) can solve the issue.

**Comment by kiwi0fruit** [comment]
At the moment the only way to get reproducible sha256 checked conda environment is to:

* create and test environment
* export it via `conda list --explicit > spec-file.txt` to text file ([see here](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#building-identical-conda-environments) for more details)
* via external script iterate through this list, find relevant files in `<CondaRoot>/pkgs` and calculate sha256 hashes and write them to `hash-file.txt`list.
* via external script run  iterate through `spec-file.txt` and download files to some directory (conda doesn't have an option to download such list without running link scripts as `--download-only` doesn't work for `--file` or direct links/paths to packages)
* check sha256 hashes (`hash-file.txt`) of the downloaded files via external script
* change this list to local paths to downloaded files (or list directory contents to a file `spec-file-local.txt`)
* install this local packages
  * create empty conda environment: `conda create -n clone`, activate it
  * conda may give you errors like:
    ```
    CondaError: Unable to determine python site-packages dir in target_prefix!
    Please make sure python is installed in C:\Python\Miniconda3\envs\test
    ```
  * hence first separately install package that gives you necessary folder structure like `./python-3.8.5-h5fd99cc_1.conda` package.
  * then install the rest: `conda install --file ./spec-file-local.txt`


### That's a rather inconvenient way to manage conda environments.

**Comment by jezdez** [comment]
For the record, this was merged in #14048.
